### ISSUE TITLE
message_codes should explicitly support Map<String, dynamic> to better support JSON (Firebase)

### ISSUE THREAD

**Issue by jmrboosties** [flutter/flutter#17417]
@kevmoo suggested i file an issue report here in regards to our discussion in this thread: https://github.com/dart-lang/json_serializable/issues/137

To summarize, Flutter's Firebase package returns documents from Firestore collections as _InternalLinkedHashMap<dynamic, dynamic>, but Flutter's recommended json package [json_serializable](https://pub.dartlang.org/packages/json_serializable) expects json to come in as a Map<String, dynamic>. This results in the following error:

```
I/flutter ( 6352): type '_InternalLinkedHashMap<dynamic, dynamic>' is not a subtype of type 'Map<String, dynamic>' where
I/flutter ( 6352):   _InternalLinkedHashMap is from dart:collection
I/flutter ( 6352):   Map is from dart:core
I/flutter ( 6352):   String is from dart:core
```

Here is an example of some code which will trigger this:

```dart
_firestore = Firestore.instance;

return _firestore.collection('favorites').where('user_id', isEqualTo: uid).getDocuments()
				.then((querySnap) {
					querySnap.documents.forEach((document) {
						print(document.data);

						Map<String, dynamic> json = document.data; //casts, but if you put breaklines through this its that _InternalLinkedHashMap<dynamic, dynamic> type
						Album album = new Album.fromJson(json['favorite_albums'][0]); //Throws the type exception
					});
				})
				.catchError((error) {
					print(error);
				});
```

Thank you!

**Comment by sbesnard** [comment]
It is a bug, in the meantime, you can use something like [this](https://github.com/flutter/flutter/issues/16589)

If you have nested Map, well, you will have to manually modify the .g.dart file to replace `cast`s by `Map<String, dynamic>.from()`

**Comment by kevmoo** [comment]
You can also use the latest json_serializable with the `any_map` option set
to true!

See the readme

**Comment by sbesnard** [comment]
Wow, took me ages to find how to use this... I had to create a build.yaml.
Any other way ? the documentation is not clear on how to use this flag with the command line.
Thanks

**Comment by kevmoo** [comment]
build.yaml with this content should be good. Did you try that?

```yaml
targets:
  $default:
    builders:
      json_serializable:
        options:
          any_map: true
```

**Comment by sbesnard** [comment]
I did, and it worked, my question was : how to pass the anyMap parameter without using a build.yaml

**Comment by jakemac53** [comment]
We do support a `--define` flag which is a bit verbose but would look something like this: `--define=json_serializable=any_map=true`

**Comment by escamoteur** [comment]
jaguar_serializer is also affected by this issue @tejainece

**Comment by escamoteur** [comment]
For jaguar_serializer this was fixed right now
jaguar_serializer: 2.1.5
jaguar_serializer_cli. 2.1.14

**Comment by stavangr** [comment]
Setting any_map to true didn't work for me. Any tips?
Currently using a workaround found here: https://github.com/flutter/flutter/issues/15007

**Comment by kevmoo** [comment]
@stavangr – are you getting the same errors? Have you upgraded the flutter plugin to latest?

**Comment by stavangr** [comment]
@kevmoo yes i was using the latest versions. Unfortunately i can't reproduce the issue anymore since i changed the project. I'll give feedback if it happens again.

**Comment by bagintz** [comment]
@kevmoo I am getting this currently if you want me to try anything in particular :-/

**Comment by kevmoo** [comment]
Set up the configuration for `any_map: true` – See details at the bottom of this page - https://pub.dartlang.org/packages/json_serializable

**Comment by bgetsug** [comment]
I am using `json_serializable` 1.5.1 and added a `build.yaml` to the root of my project as [shown above](https://github.com/flutter/flutter/issues/17417#issuecomment-393551954), but in order for things to work for _nested objects_, I had to make sure that I type hinted a simple `Map` instead of a `Map<String, dynamic>` in the `fromJson` factory constructor.

```dart
import 'package:json_annotation/json_annotation.dart';

part 'organizations.g.dart';

JsonSerializable()
class Organization {
  @JsonKey(ignore: true)
  String id;

  String name;

  List<OrgUnit> orgUnits;

  Organization(this.name);

  factory Organization.fromJson(Map json) => _$OrganizationFromJson(json);

  Map<String, dynamic> toJson() => _$OrganizationToJson(this);
}

@JsonSerializable()
class OrgUnit {
  @JsonKey(ignore: true)
  String id;

  @JsonKey(ignore: true)
  String orgId;

  String name;

  OrgUnit(this.name);

  factory OrgUnit.fromJson(Map json) => _$OrgUnitFromJson(json);

  Map<String, dynamic> toJson() => _$OrgUnitToJson(this);
}
```

generates:

```dart
// GENERATED CODE - DO NOT MODIFY BY HAND

part of 'organizations.dart';

// **************************************************************************
// JsonSerializableGenerator
// **************************************************************************

Organization _$OrganizationFromJson(Map json) {
  return Organization(json['name'] as String)
    ..orgUnits = (json['orgUnits'] as List)
        ?.map((e) => e == null ? null : OrgUnit.fromJson(e as Map))
        ?.toList();
}

Map<String, dynamic> _$OrganizationToJson(Organization instance) =>
    <String, dynamic>{'name': instance.name, 'orgUnits': instance.orgUnits};

OrgUnit _$OrgUnitFromJson(Map json) {
  return OrgUnit(json['name'] as String);
}

Map<String, dynamic> _$OrgUnitToJson(OrgUnit instance) =>
    <String, dynamic>{'name': instance.name};
```

**Comment by kevmoo** [comment]
@zoechi – I don't think we're waiting on the customer, are we? We need the flutter/firebase package to get updated...right?

**Comment by kevmoo** [comment]
> @kevmoo that was not how interpreted above conversation.
So any_map: true is just a workaround then.

True.

**Comment by kevmoo** [comment]
So work is done to make sure any `Map<dynamic, dynamic>` is converted to `Map<String, dynamic>` – **but only the root**. This means any nested maps remain unchanged, which causes this issue. So...

1) This code could be updated to try to return a `Map<String, dynamic>` if all keys are `String`. This would involve runtime checks and some fallback.

https://github.com/flutter/flutter/blob/323d1574c6cecf1163c61f8b2a4bdaa0b3f9c32f/packages/flutter/lib/src/services/message_codecs.dart#L428-L434

2) We could update this code to have a special `JsonMap` type. Then we can ensure we create `Map<String, dynamic>`

https://github.com/flutter/plugins/blob/master/packages/cloud_firestore/lib/src/firestore_message_codec.dart#L8-L19

3) We could update this code to do a deep scan of nested values to ensure every nested map is the right type of map.

https://github.com/flutter/plugins/blob/39ba555c4ccf640b2b31943d7befc9b5cba5781d/packages/cloud_firestore/lib/src/document_snapshot.dart#L34-L40

I think the 2nd option is the best idea, honestly. Then we can get rid of https://github.com/flutter/plugins/blob/39ba555c4ccf640b2b31943d7befc9b5cba5781d/packages/cloud_firestore/lib/src/document_snapshot.dart#L34 entirely

In the meantime, the `any_map` solution is probably the best path forward.

**Comment by kevmoo** [comment]
...`Map<String, dynamic>` is so common, it might be worth adding it as a special case to the core `message_codecs.dart` impl – call that option 1.5. We pay for another map type, but we avoid a lot of casting and copying everywhere else...

**Comment by kevmoo** [comment]
CC @a14n @gspencergoog @mravn-google as folks who have touched `message_codecs`.

RE adding `_valueStringMap` as No. 14, maybe?

**Comment by felipemeirelles** [comment]
So, any news on the matter? I still have the same issue with nested Maps

**Comment by kevmoo** [comment]
@gspencergoog @turnidge ? Thoughts here?

**Comment by gspencergoog** [comment]
Yes, I think adding a type for `Map<String, dynamic>` makes sense.

**Comment by kevmoo** [comment]
@gspencergoog – should we rename this issue to run w/ that request? Or open another issue?

Not sure who who'd be able to run on this. I have neither the time nor talent. 😄

**Comment by gspencergoog** [comment]
(Well, I disagree on the "talent" part. :-))

Yeah, let's rename it. I don't know when I'll be able to get to it either, but it should be taken care of. Maybe @collinjackson or @bparrishMines want to take a look?

**Comment by kevmoo** [comment]
Friendly ping. Just had another user hit this w/ json_serializable. Very hard to debug.

**Comment by bparrishMines** [comment]
> So work is done to make sure any `Map<dynamic, dynamic>` is converted to `Map<String, dynamic>` – **but only the root**. This means any nested maps remain unchanged, which causes this issue. So...
> 
> 1. This code could be updated to try to return a `Map<String, dynamic>` if all keys are `String`. This would involve runtime checks and some fallback.
> 
> [flutter/packages/flutter/lib/src/services/message_codecs.dart](https://github.com/flutter/flutter/blob/323d1574c6cecf1163c61f8b2a4bdaa0b3f9c32f/packages/flutter/lib/src/services/message_codecs.dart#L428-L434)
> 
> Lines 428 to 434 in [323d157](/flutter/flutter/commit/323d1574c6cecf1163c61f8b2a4bdaa0b3f9c32f)
> 
>  case _valueMap: 
>    final int length = readSize(buffer); 
>    result = <dynamic, dynamic>{}; 
>    for (int i = 0; i < length; i++) { 
>      result[readValue(buffer)] = readValue(buffer); 
>    } 
>    break; 
> 1. We could update this code to have a special `JsonMap` type. Then we can ensure we create `Map<String, dynamic>`
> 
> https://github.com/flutter/plugins/blob/master/packages/cloud_firestore/lib/src/firestore_message_codec.dart#L8-L19
> 
> 1. We could update this code to do a deep scan of nested values to ensure every nested map is the right type of map.
> 
> https://github.com/flutter/plugins/blob/39ba555c4ccf640b2b31943d7befc9b5cba5781d/packages/cloud_firestore/lib/src/document_snapshot.dart#L34-L40
> 
> I think the 2nd option is the best idea, honestly. Then we can get rid of https://github.com/flutter/plugins/blob/39ba555c4ccf640b2b31943d7befc9b5cba5781d/packages/cloud_firestore/lib/src/document_snapshot.dart#L34 entirely
> 
> In the meantime, the `any_map` solution is probably the best path forward.

I wonder if it would be possible to have `Map`s support implicit casting as `List`s do instead. https://www.dartlang.org/guides/language/sound-problems#runtime-errors Unfortunately, dart doesn't support overriding the assignment operator.

This option would probably be more effort for the Dart team though.

I'm on the fence whether we should support casting only `Map<String, dynamic>`. If a user was successfully able to use the code 

```
final Map<String, dynamic> result = await channel.invokeMethod('myMethod', null);
```

they would probably be confused why they would get an error for

```
final Map<int, dynamic> result = await channel.invokeMethod('myMethod', null);
```

Creating a `JsonMap` could be an option, but I'm not sure how to effectively extend a Map without creating a class with a map variable. It would also be created for the single purpose of casting nested maps.

cc @collinjackson

**Comment by marcelino-borges** [comment]
Hey guys! Any news about this issue?
I've struggling so much these days trying to get this json stuff working.
The sad part of it is that I remember using Firebase with native framework (java) and things used to be so simple and easy, no need of json... And with Flutter we have to struggle like this to parse to/from json all the time, and there is always a problem showing up. =(

Would we have, at least, an alternative solution to get an object from Firestore without facing this problem? I can't think of a non-ugly solution right now...

**Exception:**
`Unhandled Exception: type '_InternalLinkedHashMap<dynamic, dynamic>' is not a subtype of type 'Map<String, dynamic>' in type cast`

**This is my class.g.dart:**
`Contractor _$ContractorFromJson(Map<String, dynamic> json) {
  return Contractor(
    json['name'] as String,
    json['lastName'] as String,
    json['cpf'] as String,
    json['phone'] as String,
    json['email'] as String,
  )
    ..id = json['id'] as String
    ..pictureUrl = json['pictureUrl'] as String
    ..rating = (json['rating'] as num)?.toDouble()
    ..mainAddress = json['mainAddress'] == null
        ? null
        : Address.fromJson(json['mainAddress'] as Map<String, dynamic>)
    ..allAddresses = (json['allAddresses'] as List)
        ?.map((e) =>
            e == null ? null : Address.fromJson(e as Map<String, dynamic>))
        ?.toList()
    ..wallet = (json['wallet'] as num)?.toDouble()
    ..scheduledAppointments = (json['scheduledAppointments'] as List)
        ?.map((e) =>
            e == null ? null : Appointment.fromJson(e as Map<String, dynamic>))
        ?.toList()
    ..archivedAppointments = (json['archivedAppointments'] as List)
        ?.map((e) =>
            e == null ? null : Appointment.fromJson(e as Map<String, dynamic>))
        ?.toList()
    ..creditCard = json['creditCard'] == null
        ? null
        : CreditCard.fromJson(json['creditCard'] as Map<String, dynamic>);
}`

**flutter doctor:**
`Doctor summary (to see all details, run flutter doctor -v):
[√] Flutter (Channel stable, v1.7.8+hotfix.4, on Microsoft Windows [versÃ£o 10.0.17763.615], locale pt-BR)

[√] Android toolchain - develop for Android devices (Android SDK version 29.0.0)
[√] Android Studio (version 3.3)
[!] IntelliJ IDEA Community Edition (version 2019.1)
    X Flutter plugin not installed; this adds Flutter specific functionality.
[!] VS Code (version 1.35.0)
    X Flutter extension not installed; install from
      https://marketplace.visualstudio.com/items?itemName=Dart-Code.flutter
[√] Connected device (1 available)

! Doctor found issues in 2 categories.
`

**Comment by ChristBKK** [comment]
got this problem now also. Is there any fix?

**Comment by tapizquent** [comment]
Got this same exception just now coming from code generated by JsonSerializable

`Unhandled Exception: type '_InternalLinkedHashMap<dynamic, dynamic>' is not a subtype of type 'Map<String, dynamic>' in type cast`

I have identified that it is because of inner Maps. Root map converts just fine.

Has any solutions come up yet?

**Comment by fiskurgit** [comment]
This is frustrating, like a lot of developers my test project with Flutter is porting an established Android/iOS app that uses Firebase, at this point I hit a brick wall when trying to serialise some simple Json:

```
var remoteConfig = await config.fetchRemoteConfig();
var configUrls = remoteConfig.getString("configUrls");
print("Config urls: " + configUrls);
var configUrlsJson = json.decode(configUrls);
```

`[ERROR:flutter/lib/ui/ui_dart_state.cc(148)] Unhandled Exception: type '_InternalLinkedHashMap<String, dynamic>' is not a subtype of type 'String'`

**Comment by marcelino-borges** [comment]
Guys, please check again flutter oficial docs about json serializable (https://flutter.dev/docs/development/data-and-backend/json). I've made a pull request recently and it was already merged.

Look for the nested classes section, there are a few tricks worth to be tried. There is the "anyMap: true" trick and a bit more. Check it out and see if it works for you.

**Comment by fiskurgit** [comment]
As a new Flutter dev the path for lots of people is the above, everything is smooth then they hit this issue when they try and work with Json - workarounds aren't the point. After being stuck for a few of hours, going away and coming back to it again I found: `Map<String, dynamic> configUrlsJson = jsonDecode(configUrlsString);` - seems obvious now, but not to a noob in Flutter/Dart.

It might not seem a big issue for seasoned Dart devs but orientating in a new language and framework takes time, `anyMap: true` sounds so simple, but then they have to find out where to put that, build.yaml right? Cool, so where does that live? More research, then you have to run another command to rebuild with `flutter pub run build_runner build` - all of this for a new Dart/Flutter developer is bewildering. The issue is I believe over a year old, and it's tripping a lot of people up: `json.decode` should just work.

**Comment by kroikie** [comment]
@jmrboosties

This issue has been moved to https://github.com/FirebaseExtended/flutterfire/issues/833. Any further collaboration will be done there.

**Comment by tewshi** [comment]
> Wow, took me ages to find how to use this... I had to create a build.yaml.
> Any other way ? the documentation is not clear on how to use this flag with the command line.
> Thanks

after doing this, don't forget to
`flutter pub run build_runner build --delete-conflicting-outputs`
